TS-41945 Support for configuration cache (#665)
cqse/teamscale-jacoco-agent: teamscale-gradle-plugin/MIGRATION_35.md

# Migration guide for the `com.teamscale` Gradle plugin version 34 to 35

> Previously, the plugin provided a way to specify the partitions either globally for a project or on a per task level.

Reviewer: It would be great if you not only mention what the previous state was but also what the new situation looks like.

> Additionally, a `teamscaleReportUpload` was provided automatically on the root project that would lazily pick up all the reports and upload them to Teamscale.

Reviewer: ```suggestion
Additionally, a `teamscaleReportUpload` task was provided automatically on the root project that would lazily pick up all the reports and upload them to Teamscale.
```

> Due to the restrictions that are imposed by the Gradle configuration cache, we had to adjust the API of the plugin to provide compatibility with it.

Reviewer: ```suggestion
With this version, we make use of the Gradle configuration cache for a faster configuration phase. Due to the restrictions that are imposed by the Gradle configuration cache, we had to adjust the API of the plugin.
```

Make it clearer why we even want to do this.

## Usage of `Property` for all inputs
We switched all extension and task properties from plain types e.g. `boolean` to `Property<Boolean>` etc.
In newer Gradle versions this should be handled transparently.

> If you are still using an older version, you might need to replace `partition = "Something"` with `partition.set("Something")`.

Reviewer: do you know how old? otherwise it remains mysterious to the reader if they are affected or not ;-)

Author: Good point! I was in a hurry when I wrote this so didn't take the time to look up the details. Actually Groovy always supported this and Gradle 8.2 also enabled this by default for Kolin DSL. As we only support Gradle 8.4+ from now on, I just dropped that part.

For more details, please refer to the official documentation: [Understanding Properties and Providers](https://docs.gradle.org/current/userguide/properties_providers.html#understanding_properties).

## The `teamscaleReportUpload` task
The type of the upload task was renamed from `TeamscaleUploadTask` to `TeamscaleUpload`.
`teamscaleReportUpload` is no longer automatically added.
Instead, you need to define your own upload task for each partition.

```groovy
import com.teamscale.TeamscaleUpload
// ...

tasks.register("teamscaleReportUpload", TeamscaleUpload) {
    partition = "Unit Tests"
    message = "Gradle upload" // Optional
    // If you don't collect testwise coverage
    from(tasks.test) // Will upload the JUnit reports produced by the Test task

>     from(tasks.jacocoTestReport) // Will upload the report produced by the JacocoReport task

Reviewer: ```suggestion
    from(tasks.jacocoTestReport) // Will upload the coverage report produced by the JacocoReport task
```

    
    // or if you do collect testwise coverage
    from(tasks.tiaTests) // Will upload the Testwise Coverage report produced by the TestImpacted task
}
```

> TODO aggregate at top-level

Reviewer: todo left in the text

## Specifying partitions top-level

> The `teamscale.report` extension action was removed.

Reviewer: ```suggestion
The `teamscale.report` extension action was removed.
This was previously used to specify the partition to use during the report upload.
```

> The partition is now specified via the `TeamscaleUpload` task.

Reviewer: ```suggestion
The partition is now specified via the `TeamscaleUpload` task, see above.
```

When you use test impact analysis, you need to additionally replace `teamscale.report.testwiseCoverage`
```groovy
teamscale {
    report {
        testwiseCoverage {
            partition = "Unit Tests"
        }
    }
}
```
with one of the following:
If you have multiple `TestImpacted` tasks that should receive the same partition:
```groovy
import com.teamscale.TestImpacted
// ...
tasks.withType(TestImpacted) {
    partition = "Unit Tests"
}
```

> If there is only one `TestImpacted` task to which the partition should apply:

Reviewer: Can't I just also do the above in that case? That would simplify the instructions.

Additionally: why do I need to still specify the partition in the TestImpacted task? Shouldn't it be enough to specify it in the upload task? If not, then it's not clear from the text when to specify the partition in which location and why.

Reviewer: Also, why can I only do this for the TestImpacted task but not for JUnit results or Jacoco coverage? seems inconsintent

Author: Because for the TestImpacted task we also need the partition to retrieve impacted tests from Teamscale. That's why the JUnit task for example does not need this.

```groovy
import com.teamscale.TestImpacted
// ...
tasks.register("tiaUnitTests", TestImpacted) {
    partition = "Unit Tests"
}
```

## `TestwiseCoverageReportTask` was merged into `TestImpacted`

> The testwise coverage report is now generated directly by the corresponding `TestImpacted` task (same like the native Gradle reports `junitXml` and `html`).

Reviewer: ```suggestion
The testwise coverage report is now generated directly by the corresponding `TestImpacted` task (same like the native Gradle test result reports `junitXml` and `html`).
```
